python/tests/test_ownership_variable.py

```diff
@@ -0,0 +1,543 @@
+    s_deepcopy = deepcopy(s)
+    s_methcopy = s.copy(deep=False)
+    s_methdeepcopy = s.copy(deep=True)
+
+    s.value['num'] = -1
+    s.value['list'][0] = -2
+    modified = {'num': -1, 'list': [-2, 3]}
+    assert sc.identical(s, make_variable(modified))
+    assert sc.identical(s_copy, make_variable(modified))
+    assert sc.identical(s_deepcopy, make_variable(data))
+    assert sc.identical(s_methcopy, make_variable(modified))
+    assert sc.identical(s_methdeepcopy, make_variable(data))
+
+
+def test_own_var_scalar_str_get():
+    # Python strings are immutable, no references are shared.
+    s = make_variable('abc')
+    x = s.value
+    s.value = 'def'
+    assert sc.identical(s, make_variable('def'))
+    assert x == 'abc'
+
+
+def test_own_var_scalar_str_copy():
+    # Depth of copies of variables can be controlled.
+    s = make_variable('abc')
+    s_copy = copy(s)
+    s_deepcopy = deepcopy(s)
+    s_methcopy = s.copy(deep=False)
+    s_methdeepcopy = s.copy(deep=True)
+
+    s.value = 'def'
+    assert sc.identical(s, make_variable('def'))
+    assert sc.identical(s_copy, make_variable('def'))
+    assert sc.identical(s_deepcopy, make_variable('abc'))
+    assert sc.identical(s_methcopy, make_variable('def'))
+    assert sc.identical(s_methdeepcopy, make_variable('abc'))
+
+
+def test_own_var_scalar_nested_set():
+    # Variables are shared when nested.
+    a = np.arange(5)
+    inner = make_variable(a, unit='m')
+    outer = make_variable(inner)
+    outer.value['x', 0] = -1
+    outer.value.values[1] = -2
+    inner['x', 2] = -3
+    inner.unit = 's'
+    a[3] = -4
+    assert sc.identical(
+        outer, make_variable(make_variable([-1, -2, -3, 3, 4], unit='s')))
+    assert sc.identical(inner, make_variable([-1, -2, -3, 3, 4], unit='s'))
+    np.testing.assert_array_equal(a, [0, 1, 2, -4, 4])
+
+    # Assignment creates a new inner Variable.
+    outer.value = make_variable(np.arange(5, 8), unit='kg')
+    outer.value['x', 0] = -5
+    assert sc.identical(outer,
+                        make_variable(make_variable([-5, 6, 7], unit='kg')))
+    assert sc.identical(inner, make_variable([-1, -2, -3, 3, 4], unit='s'))
+
+    # TODO this causes and infinite loop and segfaults at the moment
+    # outer.value = outer
+    # assert sc.identical(outer, outer.value)
+
+
+def test_own_var_scalar_nested_get():
+    # Variables are shared when nested.
+    outer = make_variable(make_variable(np.arange(5)))
+    inner = outer.value
+    a = inner.values
+    outer.value['x', 0] = -1
+    outer.value.values[1] = -2
+    inner['x', 2] = -3
+    a[3] = -4
+    expected = [-1, -2, -3, -4, 4]
+    assert sc.identical(outer, make_variable(make_variable(expected)))
+    assert sc.identical(inner, make_variable(expected))
+    np.testing.assert_array_equal(a, expected)
+
+    # Assignment writes the values into the inner Variable.
+    outer.value = make_variable(np.arange(5, 8))
+    outer.value['x', 0] = -5
+    inner['x', 1] = -10
```

Comment:
I assume it is re-pointing the inner variable to the RHS, not copying the values? If not, shouldn't it?

Reply:
True, `a` doesn't change, so `inner` points at a different buffer. That makes it difficult to reason about unless you know about the inner workings.